File: python/db/schema.py
"""SQLAlchemy ORM models derived from src/database/db2/db2-definitions.sql.

Tables:
- PortfolioMaster   (PORTFOLIO_MASTER)
- InvestmentPositions (INVESTMENT_POSITIONS)
- TransactionHistory (TRANSACTION_HISTORY)
- ErrorLog          (ERRLOG)          — from ERRLOG.sql / ERRHNDL.cbl / DB2ERR.cbl
- AuditLog          (AUDITLOG)        — from SECMGR.cbl
- AuthFile          (AUTHFILE)        — from SECMGR.cbl authorization checks
"""

from datetime import date, datetime, time
from decimal import Decimal

from sqlalchemy import (
    CHAR,
    TIMESTAMP,
    VARCHAR,
    Date,
    ForeignKeyConstraint,
    Index,
    Integer,
    Numeric,
    Time,
)
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship


class Base(DeclarativeBase):
    """Declarative base for all ORM models."""


# -------------------------------------------------------------------------
# PORTFOLIO_MASTER
# -------------------------------------------------------------------------


class PortfolioMaster(Base):
    """Portfolio master table (db2-definitions.sql PORTFOLIO_MASTER)."""

    __tablename__ = "portfolio_master"

    portfolio_id: Mapped[str] = mapped_column(CHAR(8), primary_key=True)
    account_type: Mapped[str] = mapped_column(CHAR(2), nullable=False)
    branch_id: Mapped[str] = mapped_column(CHAR(2), nullable=False)
    client_id: Mapped[str] = mapped_column(CHAR(10), nullable=False)
    portfolio_name: Mapped[str] = mapped_column(VARCHAR(50), nullable=False)
    currency_code: Mapped[str] = mapped_column(CHAR(3), nullable=False)
    risk_level: Mapped[str] = mapped_column(CHAR(1), nullable=False)
    status: Mapped[str] = mapped_column(CHAR(1), nullable=False)
    open_date: Mapped[date] = mapped_column(Date, nullable=False)
    close_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    last_maint_date: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    last_maint_user: Mapped[str] = mapped_column(VARCHAR(8), nullable=False)

    # Relationships
    positions: Mapped[list["InvestmentPositions"]] = relationship(back_populates="portfolio")
    transactions: Mapped[list["TransactionHistory"]] = relationship(
        back_populates="portfolio"
    )

    __table_args__ = (
        Index("idx_port_master_client", "client_id", "status"),
    )


# -------------------------------------------------------------------------
# INVESTMENT_POSITIONS
# -------------------------------------------------------------------------


class InvestmentPositions(Base):
    """Investment positions table (db2-definitions.sql INVESTMENT_POSITIONS)."""

    __tablename__ = "investment_positions"

    portfolio_id: Mapped[str] = mapped_column(CHAR(8), primary_key=True)
    investment_id: Mapped[str] = mapped_column(CHAR(10), primary_key=True)
    position_date: Mapped[date] = mapped_column(Date, primary_key=True)
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    cost_basis: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    market_value: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    currency_code: Mapped[str] = mapped_column(CHAR(3), nullable=False)
    last_maint_date: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    last_maint_user: Mapped[str] = mapped_column(VARCHAR(8), nullable=False)

    # Relationship
    portfolio: Mapped["PortfolioMaster"] = relationship(back_populates="positions")

    __table_args__ = (
        ForeignKeyConstraint(["portfolio_id"], ["portfolio_master.portfolio_id"]),
        Index("idx_positions_date", "position_date", "portfolio_id"),
    )


# -------------------------------------------------------------------------
# TRANSACTION_HISTORY
# -------------------------------------------------------------------------


class TransactionHistory(Base):
    """Transaction history table (db2-definitions.sql TRANSACTION_HISTORY)."""

    __tablename__ = "transaction_history"

    transaction_id: Mapped[str] = mapped_column(CHAR(20), primary_key=True)
    portfolio_id: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    transaction_date: Mapped[date] = mapped_column(Date, nullable=False)
    transaction_time: Mapped[time] = mapped_column(Time, nullable=False)
    investment_id: Mapped[str] = mapped_column(CHAR(10), nullable=False)
    transaction_type: Mapped[str] = mapped_column(CHAR(2), nullable=False)
    quantity: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    price: Mapped[Decimal] = mapped_column(Numeric(18, 4), nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(18, 2), nullable=False)
    currency_code: Mapped[str] = mapped_column(CHAR(3), nullable=False)
    status: Mapped[str] = mapped_column(CHAR(1), nullable=False)
    process_date: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    process_user: Mapped[str] = mapped_column(VARCHAR(8), nullable=False)

    # Relationship
    portfolio: Mapped["PortfolioMaster"] = relationship(back_populates="transactions")

    __table_args__ = (
        ForeignKeyConstraint(["portfolio_id"], ["portfolio_master.portfolio_id"]),
        Index("idx_trans_hist_port", "portfolio_id", "transaction_date"),
        Index("idx_trans_hist_date", "transaction_date", "portfolio_id"),
    )


# -------------------------------------------------------------------------
# ERRLOG  (from ERRLOG.sql, used by ERRHNDL.cbl and DB2ERR.cbl)
# -------------------------------------------------------------------------


class ErrorLog(Base):
    """Error logging table (ERRLOG.sql)."""

    __tablename__ = "errlog"

    error_timestamp: Mapped[datetime] = mapped_column(TIMESTAMP, primary_key=True)
    program_id: Mapped[str] = mapped_column(CHAR(8), primary_key=True)
    error_type: Mapped[str] = mapped_column(CHAR(1), nullable=False)
    error_severity: Mapped[int] = mapped_column(Integer, nullable=False)
    error_code: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    error_message: Mapped[str] = mapped_column(VARCHAR(200), nullable=False)
    process_date: Mapped[date] = mapped_column(Date, nullable=False)
    process_time: Mapped[time] = mapped_column(Time, nullable=False)
    user_id: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    additional_info: Mapped[str | None] = mapped_column(VARCHAR(500), nullable=True)

    __table_args__ = (
        Index("errlog_ix1", "process_date", "error_severity"),
    )


# -------------------------------------------------------------------------
# AUDITLOG  (used by SECMGR.cbl for audit trail)
# -------------------------------------------------------------------------


class AuditLog(Base):
    """Audit log table used by SECMGR.cbl for security audit trails."""

    __tablename__ = "auditlog"

    audit_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    audit_timestamp: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    system_id: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    user_id: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    program: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    terminal: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    audit_type: Mapped[str] = mapped_column(CHAR(4), nullable=False)
    action: Mapped[str] = mapped_column(CHAR(8), nullable=False)
    status: Mapped[str] = mapped_column(CHAR(4), nullable=False)
    portfolio_id: Mapped[str | None] = mapped_column(CHAR(8), nullable=True)
    account_no: Mapped[str | None] = mapped_column(CHAR(10), nullable=True)
    before_image: Mapped[str | None] = mapped_column(VARCHAR(100), nullable=True)
    after_image: Mapped[str | None] = mapped_column(VARCHAR(100), nullable=True)
    message: Mapped[str | None] = mapped_column(VARCHAR(100), nullable=True)

    __table_args__ = (
        Index("idx_auditlog_user", "user_id", "audit_timestamp"),
        Index("idx_auditlog_portfolio", "portfolio_id", "audit_timestamp"),
    )


# -------------------------------------------------------------------------
# AUTHFILE  (used by SECMGR.cbl for authorization checks)
# -------------------------------------------------------------------------


class AuthFile(Base):
    """Authorization file used by SECMGR.cbl for access control."""

    __tablename__ = "authfile"

    user_id: Mapped[str] = mapped_column(CHAR(8), primary_key=True)
    password_hash: Mapped[str] = mapped_column(VARCHAR(128), nullable=False)
    user_name: Mapped[str] = mapped_column(VARCHAR(50), nullable=False)
    access_level: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(CHAR(1), nullable=False)
    last_login: Mapped[datetime | None] = mapped_column(TIMESTAMP, nullable=True)
    failed_attempts: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_date: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    last_maint_date: Mapped[datetime] = mapped_column(TIMESTAMP, nullable=False)
    last_maint_user: Mapped[str] = mapped_column(VARCHAR(8), nullable=False)

    __table_args__ = (
        Index("idx_authfile_status", "status", "user_id"),
    )

File: ts-src/database/schema.ts
/**
 * Database Schema / Migrations.
 * Migrated from: src/database/db2/db2-definitions.sql
 *
 * Creates tables equivalent to the DB2 PORTFOLIO_MASTER, INVESTMENT_POSITIONS,
 * TRANSACTION_HISTORY, RTNCODES, POSHIST, ERRLOG, and AUDITLOG tables.
 */

import { Knex } from 'knex';

/** Run all table-creation migrations. */
export async function createSchema(db: Knex): Promise<void> {
  // PORTFOLIO_MASTER
  if (!(await db.schema.hasTable('PORTFOLIO_MASTER'))) {
    await db.schema.createTable('PORTFOLIO_MASTER', (t) => {
      t.string('PORTFOLIO_ID', 8).primary();
      t.string('ACCOUNT_NO', 10).notNullable();
      t.string('ACCOUNT_TYPE', 2).defaultTo('');
      t.string('CLIENT_NAME', 30).defaultTo('');
      t.string('CLIENT_TYPE', 1).defaultTo('I');
      t.string('BRANCH_ID', 3).defaultTo('');
      t.string('STATUS', 1).notNullable().defaultTo('A');
      t.decimal('TOTAL_VALUE', 15, 2).defaultTo(0);
      t.decimal('CASH_BALANCE', 15, 2).defaultTo(0);
      t.decimal('TOTAL_COST', 15, 2).defaultTo(0);
      t.integer('TOTAL_UNITS').defaultTo(0);
      t.string('CREATE_DATE', 10).defaultTo('');
      t.string('LAST_MAINT_DATE', 10).defaultTo('');
      t.string('MAINT_USER', 8).defaultTo('');
    });
  }

  // INVESTMENT_POSITIONS
  if (!(await db.schema.hasTable('INVESTMENT_POSITIONS'))) {
    await db.schema.createTable('INVESTMENT_POSITIONS', (t) => {
      t.string('PORTFOLIO_ID', 8).notNullable();
      t.string('INVESTMENT_ID', 8).notNullable();
      t.string('POSITION_DATE', 10).notNullable();
      t.string('ACCOUNT_NO', 10).defaultTo('');
      t.string('DESCRIPTION', 30).defaultTo('');
      t.string('INVESTMENT_TYPE', 3).defaultTo('');
      t.string('STATUS', 1).defaultTo('A');
      t.decimal('QUANTITY', 15, 3).defaultTo(0);
      t.decimal('COST_BASIS', 15, 2).defaultTo(0);
      t.decimal('MARKET_VALUE', 15, 2).defaultTo(0);
      t.decimal('PERCENT_CHANGE', 7, 2).defaultTo(0);
      t.string('LAST_VAL_DATE', 10).defaultTo('');
      t.primary(['PORTFOLIO_ID', 'INVESTMENT_ID', 'POSITION_DATE']);
    });
  }

  // TRANSACTION_HISTORY
  if (!(await db.schema.hasTable('TRANSACTION_HISTORY'))) {
    await db.schema.createTable('TRANSACTION_HISTORY', (t) => {
      t.string('TRANS_DATE', 8).notNullable();
      t.string('TRANS_TIME', 6).notNullable();
      t.string('PORTFOLIO_ID', 8).notNullable();
      t.integer('SEQUENCE_NO').notNullable();
      t.string('TRANS_TYPE', 2).defaultTo('');
      t.string('STATUS', 1).defaultTo('P');
      t.string('INVESTMENT_ID', 8).defaultTo('');
      t.decimal('QUANTITY', 13, 2).defaultTo(0);
      t.decimal('PRICE', 13, 2).defaultTo(0);
      t.decimal('AMOUNT', 15, 2).defaultTo(0);
      t.decimal('FEES', 13, 2).defaultTo(0);
      t.string('ACCOUNT_NO', 10).defaultTo('');
      t.string('DESCRIPTION', 30).defaultTo('');
      t.primary(['TRANS_DATE', 'TRANS_TIME', 'PORTFOLIO_ID', 'SEQUENCE_NO']);
    });
  }

  // POSHIST (Position History – from POSHIST.sql)
  if (!(await db.schema.hasTable('POSHIST'))) {
    await db.schema.createTable('POSHIST', (t) => {
      t.string('ACCOUNT_NO', 10).notNullable();
      t.string('PORTFOLIO_ID', 8).notNullable();
      t.string('TRANS_DATE', 10).notNullable();
      t.string('TRANS_TIME', 8).notNullable();
      t.string('TRANS_TYPE', 4).defaultTo('');
      t.string('SECURITY_ID', 8).defaultTo('');
      t.decimal('QUANTITY', 15, 3).defaultTo(0);
      t.decimal('PRICE', 15, 3).defaultTo(0);
      t.decimal('AMOUNT', 15, 2).defaultTo(0);
      t.decimal('FEES', 15, 2).defaultTo(0);
      t.decimal('TOTAL_AMOUNT', 15, 2).defaultTo(0);
      t.decimal('COST_BASIS', 15, 2).defaultTo(0);
      t.decimal('GAIN_LOSS', 15, 2).defaultTo(0);
      t.string('PROCESS_DATE', 10).defaultTo('');
      t.string('PROCESS_TIME', 8).defaultTo('');
      t.string('USER_ID', 8).defaultTo('');
      t.primary(['ACCOUNT_NO', 'PORTFOLIO_ID', 'TRANS_DATE', 'TRANS_TIME']);
    });
  }

  // ERRLOG (Error Log – from ERRLOG.sql)
  if (!(await db.schema.hasTable('ERRLOG'))) {
    await db.schema.createTable('ERRLOG', (t) => {
      t.string('ERROR_TIMESTAMP', 26).notNullable();
      t.string('PROGRAM_ID', 8).notNullable();
      t.string('ERROR_TYPE', 1).defaultTo('S');
      t.integer('ERROR_SEVERITY').defaultTo(1);
      t.string('ERROR_CODE', 8).defaultTo('');
      t.string('ERROR_MESSAGE', 80).defaultTo('');
      t.string('PROCESS_DATE', 10).defaultTo('');
      t.string('PROCESS_TIME', 8).defaultTo('');
      t.string('USER_ID', 8).defaultTo('');
      t.string('ADDITIONAL_INFO', 100).defaultTo('');
      t.primary(['ERROR_TIMESTAMP', 'PROGRAM_ID']);
    });
  }

  // RTNCODES (Return Code Logging – from RTNCODES.sql)
  if (!(await db.schema.hasTable('RTNCODES'))) {
    await db.schema.createTable('RTNCODES', (t) => {
      t.string('TIMESTAMP', 26).notNullable();
      t.string('PROGRAM_ID', 8).notNullable();
      t.integer('RETURN_CODE').notNullable();
      t.integer('HIGHEST_CODE').notNullable();
      t.string('STATUS_CODE', 1).notNullable();
      t.string('MESSAGE_TEXT', 80).defaultTo('');
      t.primary(['TIMESTAMP', 'PROGRAM_ID']);
    });
  }

  // AUDITLOG
  if (!(await db.schema.hasTable('AUDITLOG'))) {
    await db.schema.createTable('AUDITLOG', (t) => {
      t.string('TIMESTAMP', 26).notNullable();
      t.string('USER_ID', 8).defaultTo('');
      t.string('TERMINAL_ID', 8).defaultTo('');
      t.string('TRANS_ID', 4).defaultTo('');
      t.string('PROGRAM', 8).defaultTo('');
      t.string('ACCESS_TYPE', 8).defaultTo('');
      t.string('TYPE', 4).defaultTo('');
      t.string('ACTION', 8).defaultTo('');
      t.string('STATUS', 4).defaultTo('');
      t.string('PORTFOLIO_ID', 8).defaultTo('');
      t.string('ACCOUNT_NO', 10).defaultTo('');
      t.string('BEFORE_IMAGE', 100).defaultTo('');
      t.string('AFTER_IMAGE', 100).defaultTo('');
      t.string('MESSAGE', 100).defaultTo('');
    });
  }

  // Create indexes equivalent to those in db2-definitions.sql
  try {
    await db.schema.table('PORTFOLIO_MASTER', (t) => {
      t.index(['ACCOUNT_NO'], 'IDX_PORT_ACCT');
      t.index(['CREATE_DATE'], 'IDX_PORT_DATE');
    });
  } catch {
    // Indexes may already exist
  }

  try {
    await db.schema.table('INVESTMENT_POSITIONS', (t) => {
      t.index(['POSITION_DATE'], 'IDX_POS_DATE');
    });
  } catch {
    // Indexes may already exist
  }

  try {
    await db.schema.table('RTNCODES', (t) => {
      t.index(['PROGRAM_ID', 'TIMESTAMP'], 'RTNCODES_PRG_IDX');
      t.index(['STATUS_CODE', 'TIMESTAMP'], 'RTNCODES_STS_IDX');
    });
  } catch {
    // Indexes may already exist
  }
}
